vortex-array/src/variants.rs

```diff
@@ -7,126 +7,195 @@ use std::sync::Arc;
 
 use vortex_dtype::field::Field;
 use vortex_dtype::{DType, ExtDType, FieldNames, PType};
-use vortex_error::{vortex_panic, VortexExpect as _, VortexResult};
+use vortex_error::{vortex_panic, VortexError, VortexExpect as _, VortexResult};
 
-use crate::{ArrayData, ArrayTrait};
+use crate::encoding::Encoding;
+use crate::{ArrayDType, ArrayData, ArrayTrait};
 
-pub trait ArrayVariants {
-    fn as_null_array(&self) -> Option<&dyn NullArrayTrait> {
+/// An Array encoding must declare which DTypes it can be downcast into.
+pub trait VariantsVTable<Array> {
+    fn as_null_array<'a>(&self, _array: &'a Array) -> Option<&'a dyn NullArrayTrait> {
         None
     }
 
-    fn as_null_array_unchecked(&self) -> &dyn NullArrayTrait {
-        self.as_null_array().vortex_expect("Expected NullArray")
+    fn as_bool_array<'a>(&self, _array: &'a Array) -> Option<&'a dyn BoolArrayTrait> {
+        None
```

Comment:
following our Slack conversation, why can't we just have a test here? and return `None` if it doesn't match.

Reply:
This vtable is invoked from ArrayData::as_bool_array, which does have the check: https://github.com/spiraldb/vortex/pull/1501/files#diff-48a1353d8b676587abd621c284d5e6bc771f362516d15cb2bb510b6b3c51b2acR165-R167

But if you hold a ChunkedArray, then too late, you already have an object that implements all the variant types. This is the benefit of the vtable stuff for us, we can implement behavior on an array that users cannot access via the array, they must go via some entry point trait / function that can perform validation before delegating to the encoding's vtable.